dash/include/dash/Coarray.h

```diff
@@ -0,0 +1,771 @@
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline operator value_type() const {
+    return *(_storage.lbegin());
+  }
+  
+  /**
+   * convert scalar Coarray to a global reference.
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  explicit inline operator reference() {
+    return static_cast<reference>(
+            *(_storage.begin()+static_cast<index_type>(dash::myid())));
+  }
+  
+  /**
+   * Get a reference to a member of a certain type at the
+   * specified offset
+   */
+  template<
+    typename MEMTYPE,
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline MEMTYPE & member(size_t offs) {
+    char * s_begin = reinterpret_cast<char *>(_storage.lbegin());
+    s_begin += offs;
+    return *(reinterpret_cast<MEMTYPE*>(s_begin));
+  }
+
+  /**
+   * Get the member via pointer to member
+   * \code
+   * struct value_t {double a; int b;};
+   * dash::Coarray<value_t> x;
+   * int b = x.member(&value_t::b);
+   * \endcode
+   */
+  template<
+    class MEMTYPE,
+    class P=T,
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline MEMTYPE & member(const MEMTYPE P::*mem) {
+    size_t offs = (size_t) &( reinterpret_cast<P*>(0)->*mem);
+    return member<MEMTYPE>(offs);
+  }
+
+  /**
+   * allows fortran like local access of scalars
+   * \code
+   *   dash::Coarray<int> i;
+   *   i = 42;
+   *   i += 21;
+   * \endcode
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline value_type & operator +=(const value_type & value) {
+    return *(_storage.lbegin()) += value;
+  }
+  
+  /**
+   * allows fortran like local access of scalars
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline value_type & operator -=(const value_type & value) {
+    return *(_storage.lbegin()) -= value;
+  }
+  
+  /**
+   * allows fortran like local access of scalars
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline value_type & operator *=(const value_type & value) {
+    return *(_storage.lbegin()) *= value;
+  }
+  
+  /**
+   * allows fortran like local access of scalars
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline value_type & operator /=(const value_type & value) {
+    return *(_storage.lbegin()) /= value;
+  }
+  
+  /**
+   * allows fortran like local access of scalars
+   * \code
+   *   dash::Coarray<int> i;
+   *   i = 42;
+   *   int b = i + 21;
+   * \endcode
+   */
+  template<
+    int __rank = _rank::value,
+    typename = typename std::enable_if<(__rank == 0)>::type>
+  inline value_type operator +(const value_type & value) const {
```

Comment:
Isn't that covered through the cast `operator value_type()` already?

Reply:
Unfortunately not, as C++ does not convert types over two edges. This is required for use cases like that:
```cpp
dash::Coarray<int> x;
// does not work without these operators
int a = x + 5;
// works without
int a = 5 + x;
```